src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -1048,6 +1048,112 @@ public static void M(object o)
                 );
         }
 
+        [Fact]
+        public void UnderscoreDeclaredAndDiscardPattern_01()
+        {
+            var source =
+@"class Program0
+{
+    static int Main() => 0;
+    private const int _ = 1;
+    bool M1(object o) => o is _;                             // error: cannot use _ as a constant
+    bool M2(object o) => o switch { 1 => true, _ => false }; // error: _ in scope
+}
+class Program1
+{
+    class _ {}
+    bool M1(object o) => o is _;                             // error: is type named _
+    bool M2(object o) => o switch { 1 => true, _ => false }; // error: _ in scope
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (11,31): error CS8413: The name '_' refers to the type 'Program1._', not the discard pattern. Use '@_' for the type, or 'var _' to discard.
+                //     bool M1(object o) => o is _;                             // error: is type named _
+                Diagnostic(ErrorCode.WRN_IsTypeNamedUnderscore, "_").WithArguments("Program1._").WithLocation(11, 31),
+                // (5,31): error CS8412: A constant named '_' cannot be used as a pattern.
+                //     bool M1(object o) => o is _;                             // error: cannot use _ as a constant
+                Diagnostic(ErrorCode.ERR_ConstantPatternNamedUnderscore, "_").WithLocation(5, 31),
+                // (12,48): error CS8411: The discard pattern '_' cannot be used where 'Program1._' is in scope.
+                //     bool M2(object o) => o switch { 1 => true, _ => false }; // error: _ in scope
+                Diagnostic(ErrorCode.ERR_UnderscoreDeclaredAndDiscardPattern, "_").WithArguments("Program1._").WithLocation(12, 48),
+                // (6,48): error CS8411: The discard pattern '_' cannot be used where 'Program0._' is in scope.
+                //     bool M2(object o) => o switch { 1 => true, _ => false }; // error: _ in scope
+                Diagnostic(ErrorCode.ERR_UnderscoreDeclaredAndDiscardPattern, "_").WithArguments("Program0._").WithLocation(6, 48)
+                );
+        }
```

Comment:
>} [](start = 8, length = 1)

Perhaps test `o is _` when a symbol `_` is in scope that is neither a constant nor a type.

Reply:
Already done in `UnderscoreDeclaredAndDiscardPattern_03`.

---
In reply to: [178321208](https://github.com/dotnet/roslyn/pull/25779#discussion_r178321208) [](ancestors = 178321208)